### PR TITLE
!feat: Add a new `evaluate_model`,  `gradient_descent` and `compute_statistics` functions

### PR DESCRIPTION
**Pull request by TheoRudkiewicz**
This PR introduce 3 functions:
- `evaluate_model` to evaluate the model with our without extension
- `gradient_descent` to perform stochastic gradient descent
- `compute_statistics` to perform the statistics computation by iterative forward/backward passes

To do:
- [x] think about `train` name
- [x] make evaluate compatible with current `extended_forward` type
- [x] merge with `utils.py` (`evaluate_extended_dataset`,  `evaluate_dataset``)
-  [x] write unit tests

Migration guide:
Instead of:
`from gromo.utils.utils import evaluate_extended_dataset, evaluate_dataset`
use 
`from gromo.utils.trainning_utils import evaluate_extended_dataset, evaluate_dataset`

I could not leave the function at their original spot due to circular imports.

### CODE REVIEW

File: src/gromo/utils/trainning_utils.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+        n : int, optional
+            The number of samples that `val` represents. Default is 1.
+        """
+        if val.item() != torch.nan and val.item() != torch.inf:
+            if self.sum is None:
+                self.sum = val * n
+            else:
+                self.sum += val * n
+            self.count += n
+
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/utils/trainning_utils.py]: Done.

File: src/gromo/utils/training_utils.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
+from collections.abc import Callable, Generator
+from typing import Any
+
+import torch
+import torch.utils.data
+from torch import nn
+from torchmetrics import Metric, classification
+
+from gromo.containers.growing_container import GrowingContainer, GrowingModel
+from gromo.utils.utils import global_device
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/utils/training_utils.py]: Fixed.

File: src/gromo/utils/training_utils.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+# backward compatibility
+# I could not keep it in utils.py because of circular imports,
+# with `global_device` being defined in utils.py and used
+# in `growing_container.py`
+def evaluate_extended_dataset(
+    model: nn.Module,
+    dataloader: torch.utils.data.DataLoader,
+    loss_fn: Callable,
+    mask: dict = {},
+) -> tuple[float, float]:
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/utils/training_utils.py]: Done.

File: src/gromo/utils/training_utils.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+            )
+    if batch_limit is None:
+        if epochs is None:
+            batch_limit = len(dataloader) + 1
+        else:
+            batch_limit = int(len(dataloader) * epochs)
+    for i, batch in enumerate(dataloader):
+        if i >= batch_limit:
+            break
+        yield i, batch
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/utils/training_utils.py]: Fixed.

File: src/gromo/utils/training_utils.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+    device = global_device()
+    _, y = next(iter(dataloader))
+    if y.dim() == 1:
+        nb_classes = model.out_features
+    else:
+        nb_classes = None
+    metric = None
+    if nb_classes is not None:
+        metric = classification.MulticlassAccuracy(model.out_features, average="micro")
+    loss, accuracy = evaluate_model(
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/utils/training_utils.py]: Fixed.

File: tests/test_training_utils.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+    def test_seed_with_generator(self):
+        """Seed is applied when dataloader has a Generator."""
+        dl = self._make_dataloader(with_generator=True)
+        batches = list(enumerate_dataloader(dl, dataloader_seed=0))
+        self.assertGreater(len(batches), 0)
+        batches_again = list(enumerate_dataloader(dl, dataloader_seed=0))
+        for (_, (x_1, y_1)), (_, (x_2, y_2)) in zip(batches, batches_again):
+            self.assertTrue(torch.equal(x_1, x_2))
+            self.assertTrue(torch.equal(y_1, y_2))
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply tests/test_training_utils.py]: Fixed.
